### PR TITLE
More like isearch

### PR DESCRIPTION
**Pull request by andyleejordan**
This implements most of the things I felt were missing in #47:

- `isearch-forward-symbol-at-point`
- `isearch-occur`
- isearch-like `M-s` bindings for those and existing ctrlf functions

The notable exception is that I haven't implemented `C-w` to yank the word at point into the search. This also needs changelog updates (and tests?) which I can provide if this is taken.

### CODE REVIEW

**Comment by andyleejordan** [comment]
~~TODO: Move to beginning of thing-at-point so search starts with thing instead of going to next thing if point is in the middle of thing.~~

**Review by raxod502 — COMMENTED** [review]
I agree, it makes more sense to reuse the `M-s` map as opposed to introduce something new like `C-o`. The only reason I added `C-o` was that I forgot about `M-s`.

Anyway, yes, having these bindings behave like they do in Isearch is a great improvement. The idea looks good to me.

**Comment by raxod502** [comment]
For symbol search, you could always use the regexp style and add delimiters.

> Move to beginning of thing-at-point so search starts with thing instead of going to next thing if point is in the middle of thing

Just a warning, this may be a bit tricky. Currently CTRLF just does a standard `search-forward` or `search-backward` from point. Does Isearch do something fancier?

**Comment by andyleejordan** [comment]
Thanks @raxod502! I can split out the easy stuff (the bindings to existing functions, namely) while working on implementing the harder stuff (forward symbol search and occur). Any other tips/suggestions would be very welcome as I haven't worked a ton with search interfaces before (having to change point back from minibuffer to original buffer was one I learned doing this last night).

**Comment by andyleejordan** [comment]
> > Move to beginning of thing-at-point so search starts with thing instead of going to next thing if point is in the middle of thing
> 
> Just a warning, this may be a bit tricky. Currently CTRLF just does a standard `search-forward` or `search-backward` from point. Does Isearch do something fancier?

This ended up being a bit easier than I think we both expected. The function `isearch-symbol-at-point` cleverly uses `find-tag-default-bounds` to position the point at the beginning of the symbol at point. I made sure that that we retained the user's actual starting position so that `ctrlf-cancel` works appropriately.

I also deleted `ctrlf-backward-symbol-at-point` as there is no isearch equivalent, so that's inconsistent, and maintaining two is annoying.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el:120]: Do you want all of these realigned on the rightmost `.` now?

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: Should we keep this and the other `C-o` binding?

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el:151]: I wish I could actually just remap these, but they are only temporarily bound when an isearch is active.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: If you think of better variable names, let me know.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: I'm very new to manipulating the minibuffer contents like this. This was the best I could find based on research into `previous-history-element`. I've caught myself accidentally inserting over the entire current buffer when my code didn't properly leave the `with-current-buffer` statement before calling `ctrlf-forward`. This would be more of a crutch, and I think correct code should not require it.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: I'm wondering if I should just pass `pos` through here too, but it doesn't make much other sense for `ctrlf-forward`. There are two main scenarios where we'd call `ctrlf-forward` from `ctrlf-forward-symbol-at-point`.

1. Hit `C-s`, enter no input, realize you want symbol under point, hit `M-s .` to start searching for it.
2. Be in the middle of a search, partially within a symbol, desire to search for the rest of the symbol, hit `M-s .` to fill it in. (Kind of like `C-w`, which I haven't implemented yet, but for the full symbol.)

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: I can revert these changes since I removed `ctrlf-backward-symbol-at-point`.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: Is there a pattern elisp pattern to follow for this? I need to get the values of `arg` and `pos` as they are within the `with-current-buffer` scope, but  I need to use them outside of it.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: Sorry, this isn't finished yet at all. It works for the very basic cases though!

**Comment by andyleejordan** [comment]
I have much improved `ctrlf-forward-symbol-at-point`, now to improve `ctrlf-occur` and perhaps implement `isearch-yank-word-or-char`.

**Comment by andyleejordan** [comment]
@raxod502 Why is `(field-string (point-max))` used instead of `(minibuffer-contents)` in places? I saw the former after using the latter myself and am curious if there's a particular reason.

**Comment by andyleejordan** [comment]
So, `ctrlf-occur` ended up being super easy to do. I think this is ready for review (except changelog updates). I'd love to add `C-w` too, but that one seems tricky to me, especially to get the isearch semantics right.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el:446]: This looks like way more change than it actually is because of indentation. I just added handling of `ctrlf--minibuffer` being `nil` (and ensured it is such if a ctrlf search isn't active) such that the message is displayed consistently as if `minibuffer-message` were used directly. This lets us use `ctrlf--message` even if a search _wasn't_ started (like no symbol at point was found).

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el:477]: And then I had to add a newline to satisfy linters.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el:838]: This is that cleanup. Maybe this belongs in the `ctrlf--finalize`? I wasn't entirely sure. We need this because it's insufficient to check `ctrlf--active-p` if some logic is being performed before the start of a search.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: Just calling out this question again 😂

Comment (andyleejordan) [reply ctrlf.el]: If given `arg` I just want this function to replace any existing input with it.

**Comment by andyleejordan** [comment]
You know, I had an idea last night that will add another isearch-like feature and simplify some things here. If I add a handler `symbol` we can have `M-s _` for `ctrlf-forward-symbol`, and defer to this function inside the symbol at point version.

**Comment by andyleejordan** [comment]
Well, that was easy.

**Comment by andyleejordan** [comment]
Okay cool. I'm happy with this now. It's missing readme and changelog updates, but I'll do those after a review.

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el:446]: (Try `M-s .` with no symbol at point, then cancel and try `C-s` followed by `M-s .` for a demo of the logic.)

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el]: It turned out it _was_ better to pass both `arg` and `pos` here. Since we pass `pos` straight to `ctrlf--start`, wrapper functions don't have to deal with the edge cases of being in or out of a search, which keeps this logic in exactly one place.

File: ctrlf.el
Comment (andyleejordan) [inline ctrlf.el:246]: Super nice side effect of this is that the symbol boundary constructs don't appear in the user's input, making it exactly like `isearch-forward-symbol`. Furthermore, we automatically get correct behavior with `ctrlf-occur`, too.

File: ctrlf.el
Comment (raxod502) [inline ctrlf.el]:
```suggestion
                             :case-fold ctrlf-no-uppercase-literal-p)))
```

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el:120]: It's up to your taste :)

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el]: Sure, for now. We'll update the README to not document the old binding, and remove it eventually. But there's no reason to break backwards compatibility unless needed.

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el:151]: Hmm... when I type `M-s C-h`, it says these commands are bound globally. Did I miss something?

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el:246]: That looks great.

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el:477]: Sounds good.

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el:446]: Works great!

**Comment by andyleejordan** [comment]
For what it's worth, I've been using this the last few days and it's working great!

**Comment by andyleejordan** [comment]
I've rebased, updated the readme and changelog, and on closer inspection, added `ctrlf-toggle-regexp` and `ctrlf-toggle-symbol` with matching bindings to get as close to [Isearch](https://www.gnu.org/software/emacs/manual/html_node/emacs/Special-Isearch.html) as possible.

**Comment by andyleejordan** [comment]
I want to add `isearch-query-replace`/`M-%`, `isearch-yank-word-or-char`/`C-w`, and `isearch-yank-line`/`M-s C-e` at some point too (but like, after this PR).

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el:838]: I think where you have it is fine. `ctrlf--finalize` is somewhat more of a hack, because it runs from `post-command-hook` and then removes itself right away. Ideally we wouldn't have to put anything there.

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el]: What about INITIAL-CONTENTS and POSITION, following `read-from-minibuffer`?

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el]: Okay!

File: ctrlf.el
Comment (raxod502) [reply ctrlf.el]: Maybe, but `delete-minibuffer-contents` followed by `insert` looks pretty great to me.

File: ctrlf.el
Comment (raxod502) [inline ctrlf.el:1049]: Having separate toggles for these is a great idea. That is faster to use than always having to select from a list.

File: ctrlf.el
Comment (raxod502) [inline ctrlf.el]: Hmmm... I was going to say we should avoid using internal functions in these wrappers, but then I looked at the other ones and they do that too. I guess at some point I forgot that I was trying to do that. Room for future improvement!

File: ctrlf.el
Comment (raxod502) [inline ctrlf.el]: Would it work if in `ctrlf--start` you set `ctrlf--current-starting-point` to `pos` instead of `ctrlf--starting-point`? Then you might not have to move point in this function, and hence not have to save its original value.

File: CHANGELOG.md
Comment (raxod502) [inline CHANGELOG.md]: Perhaps this would work better in the Features section (can be added just above Enhancements)?

**Review by raxod502 — COMMENTED** [review]
Sorry for the delay. This is a really great contribution, thank you for putting so much thought into it! I have some thoughts, let me know what you think. All minor things.

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el:151]: So, `M-s o` is bound globally to `occur`. But in an `isearch`, `M-s o` is transiently rebound to `isearch-occur`. An `M-s C-h` here is misleading because even if you've started an isearch, it pulls up the global bindings.

I found `which-key` really helpful here because it showed the transient bindings, and I otherwise referred to the [manual](https://www.gnu.org/software/emacs/manual/html_node/emacs/Special-Isearch.html).

However, `ctrlf-forward-symbol-at-point` I goofed on. We're already defining it globally (as does isearch) and it doesn't need to be rebound again here. The rest of these, however, are only bound during `isearch`.

Note that like `M-s _` is slightly tricky because outside of the search it should be `ctrlf-forward-symbol` but inside a search it should be `ctrlf-toggle-symbol`.

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el]: Good catch, thanks!

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el:120]: Then nah cause cause it'll be a bigger diff and things will really slide to the right.

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el:246]: Thanks! It really cleaned up when I did this.

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el:446]: Awesome.

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el:838]: Good to know.

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el]: Cool!

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el:1049]: Hey I didn't think of it! Thank you long-forgotten Emacs/Isearch developer 😆

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el]: I feel like I tried that, but let me try it out again.

**Comment by andyleejordan** [comment]
Rebased 😂 I just need to try the different approach to `pos`, and then rename the vars.

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el]: Ah ha, that worked nicely!

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el]: _Much better._ Thanks!

File: ctrlf.el
Comment (andyleejordan) [reply ctrlf.el]: Oh yeah, always room for improvement 😂

**Comment by andyleejordan** [comment]
@raxod502 I think you can merge this now when you see fit. Let me know if there's anything else!

**Comment by andyleejordan** [comment]
@raxod502 were you waiting on anything else here?

**Comment by raxod502** [comment]
Nope, I've just been very busy with other life things and have 40 threads backlogged at the moment. I will definitely get back to this as soon as possible. Sorry for the inconvenience!

**Comment by andyleejordan** [comment]
Fair! I definitely get that. Take care!

**Review by raxod502 — APPROVED** [review]
Great! Thanks for your patience!

**Pull request merged**
